cmds/switch_root/switch_root.go

```diff
@@ -23,140 +26,171 @@ func usage() string {
-
-			if fileName == "." || fileName == ".." {
-				continue
-			}
-
-			if err := littleDoctor(filepath.Join(path, fileName), fsType); err != nil {
-				return err
-			}
-			if err := os.Remove(path); err != nil {
-				return err
-			}
+		// don't descent into other file systems
+		if childFdDev, err := getDev(childFd); err != nil {
+			return err
+		} else if childFdDev != parentDev {
+			// this means continue in recursiveDelete
+			return nil
 		}
 
-	} else {
-		if err := os.Remove(path); err != nil {
+		if err:= recursiveDelete(childFd); err != nil {
+			return err
+		}
+		// back from recursion, dir is now empty, delete
+		if err := unix.Unlinkat(parentFd, childName, unix.AT_REMOVEDIR); err != nil {
 			return err
 		}
 	}
-
 	return nil
 }
 
-// execCommand will run the executable at "path" with PID 1.
+// do a proper exec to retain PID 1.
 // It returns an error if the command exits incorrectly.
 func execCommand(path string) error {
-	cmd := exec.Command(path)
-	cmd.Stdin = os.Stdin
-	cmd.Stderr = os.Stderr
-	cmd.Stdout = os.Stdout
-	cmd.SysProcAttr = &syscall.SysProcAttr{
-		Setctty:    true,
-		Setsid:     true,
-		Cloneflags: syscall.CLONE_THREAD | syscall.CLONE_NEWPID,
-	}
-
-	return cmd.Run()
+	return syscall.Exec(path, []string{path}, []string{})
 }
 
-// specialFS creates and mounts proc, sys and dev at the root level.
-func specialFS() error {
-
-	if err := syscall.Mkdir("/proc", 0); err != nil {
-		return err
-	}
-
-	if err := syscall.Mkdir("/sys", 0); err != nil {
-		return err
+// check if dir is empty
+func isEmpty(name string) (bool, error) {
+	f, err := os.Open(name)
+	if err != nil {
+		return false, err
 	}
+	defer f.Close()
 
-	if err := syscall.Mkdir("/dev", 0); err != nil {
-		return err
+	_, err = f.Readdirnames(1)
+	if err == io.EOF { // this means emtpy
+		return true, nil
 	}
+	return false, err
+}
 
-	if err := syscall.Mount("", "/proc", "proc", 0, ""); err != nil {
-		return err
-	}
+// moves a mountpoint
+func moveMount(oldPath string, newPath string) error {
+	return syscall.Mount(oldPath, newPath, "", syscall.MS_MOVE, "")
+}
 
-	if err := syscall.Mount("", "/sys", "sysfs", 0, ""); err != nil {
-		return err
+// moves common 'special' mounts if the exist
+func specialFS(newRoot string) error {
+	var mounts = []string{"/dev", "/proc", "/sys", "/run"}
+
+	for _, mount := range mounts {
+		path := filepath.Join(newRoot, mount)
+		// skip all mounting if the dir does not exists
+		if _, err := os.Stat(mount); os.IsNotExist(err) {
+			fmt.Println("switch_root: Skipping", mount)
+			continue
+		} else if err != nil {
+			return err
+		}
+		// make sure the target dir exist and is empty
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			if err := syscall.Mkdir(path, 0); err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		}
+		if empty, err := isEmpty(path); err != nil {
+			return err
+		} else if !empty {
+			return fmt.Errorf("%v must be empty", path)
```

Comment:
Since isEmpty is used to generate an error, why not just return only an error from isEmpty?